src/qml/qmlwaveformrenderer.h

```diff
@@ -0,0 +1,586 @@
+        m_textColor = value;
+    }
+    QString align() const {
+        return m_align;
+    }
+    void setAlign(const QString& value) {
+        m_align = value;
+    }
+    QString text() const {
+        return m_text;
+    }
+    void setText(const QString& value) {
+        m_text = value;
+    }
+    QString pixmap() const {
+        return m_pixmap;
+    }
+    void setPixmap(const QString& value) {
+        m_pixmap = value;
+    }
+    QString icon() const {
+        return m_icon;
+    }
+    void setIcon(const QString& value) {
+        m_icon = value;
+    }
+
+  private:
+    QString m_control;
+    QString m_visibilityControl;
+    QString m_color;
+    QString m_textColor;
+    QString m_align;
+    QString m_text;
+    QString m_pixmap;
+    QString m_icon;
+};
+
+class QmlWaveformUntilMark : public QObject {
+    Q_OBJECT
+    Q_PROPERTY(bool showTime READ showTime WRITE setShowTime NOTIFY showTimeChanged)
+    Q_PROPERTY(bool showBeats READ showBeats WRITE setShowBeats NOTIFY showBeatsChanged)
+    Q_PROPERTY(HAlignment align READ align WRITE setAlign NOTIFY alignChanged)
+    Q_PROPERTY(int textSize READ textSize WRITE setTextSize NOTIFY textSizeChanged)
+
+    QML_NAMED_ELEMENT(WaveformUntilMark)
+  public:
+    enum HAlignment { AlignTop = Qt::AlignTop,
+        AlignCenter = Qt::AlignCenter,
+        AlignBottom = Qt::AlignBottom };
+    Q_ENUM(HAlignment)
+
+    bool showTime() const {
+        return m_showTime;
+    }
+    void setShowTime(bool showTime) {
+        m_showTime = showTime;
+    }
+    bool showBeats() const {
+        return m_showBeats;
+    }
+    void setShowBeats(bool showBeats) {
+        m_showBeats = showBeats;
+    }
+    HAlignment align() const {
+        return m_align;
+    }
+    void setAlign(HAlignment align) {
+        m_align = align;
+    }
+    int textSize() const {
+        return m_textSize;
+    }
+    void setTextSize(int textSize) {
+        m_textSize = textSize;
+    }
+    float textHeightLimit() const {
+        return m_textSize;
+    }
+    void setTextHeightLimit(float textHeightLimit) {
+        m_textHeightLimit = textHeightLimit;
+    }
+  signals:
+    void showTimeChanged(bool);
+    void showBeatsChanged(bool);
+    void alignChanged(Qt::Alignment);
+    void textSizeChanged(int);
+
+  private:
+    bool m_showTime;
+    bool m_showBeats;
+    HAlignment m_align;
+    int m_textSize;
+    int m_textHeightLimit;
+};
+
+class QmlWaveformRendererMarkRange
+        : public QmlWaveformRendererFactory {
+    Q_OBJECT
+    Q_PROPERTY(QQmlListProperty<QmlWaveformMarkRange> ranges READ ranges)
+    Q_CLASSINFO("DefaultProperty", "ranges")
+    QML_NAMED_ELEMENT(WaveformRendererMarkRange)
+
+  public:
+    QmlWaveformRendererMarkRange();
+
+    QQmlListProperty<QmlWaveformMarkRange> ranges() {
+        return {this, &m_ranges};
+    }
+
+    Renderer create(WaveformWidgetRenderer* waveformWidget) const override;
+
+  private:
+    QList<QmlWaveformMarkRange*> m_ranges;
```

Comment:
what are the lifetime requirements on the `QmlWaveformMarkRange*`? Is it possible to get a use after free if Qml holds on to this list after a new track has been loaded?